### ISSUE TITLE
Calculate dipole (or other moments) from atomic charges & positions?

### ISSUE THREAD

**Issue by ghutchis** [cclib/cclib#311]
Should there be a calculation method to estimate the dipole moment from atomic coordinates and charges when available? I notice, for example that NWChem (and possibly other programs) don't provide moments for every output file.

It's simple code + using the unit converter.

**Comment by berquist** [comment]
This would be a good method to have. Here is a one-liner for the nuclear contribution that I found in some other code of mine:

    return sum(self.charges[i] * (self.coords[i] - origin) for i in range(self.size))

and this could definitely be changed to read from Mulliken charges or something similar.

(The true electronic contribution is also very simple, but we currently don't parse AO integrals other than the overlap. But having a method would make adding that later very convenient.)

**Comment by ghutchis** [comment]
You'll need to convert:
        "e.ang_to_Debye":     lambda x: x * 4.80320425,

**Comment by ghutchis** [comment]
I'd also think the method should check for the self.moments attribute and return it if present. Do you agree?

**Comment by berquist** [comment]
> I'd also think the method should check for the self.moments attribute and return it if present. Do you agree?

Yes. Later we can add an option to force (re)calculating the moments even if they're present in the output.

**Comment by langner** [comment]
Pushing back, but this seems very simple to do.

**Comment by gaursagar** [comment]
Hi, I want to help with this issue.
I have the following concerns for dipole moment calculations taking the center of mass as origin:
1. In the case of geometry optimizations, should only the optimized coordinates `self.atomcoords[-1]` should be used.
2. Which charges should be used; `mulliken` or `lowdin` have no direct mapping with electronic charges
3. Which atomic masses should be considered to calculate the center of mass in cases `self.atommasses` is not present.

In the attached code, I have added a new class `Moment` similar to those present in[/cclib/method/](https://github.com/cclib/cclib/tree/master/src/cclib/method) directory, and have used `self.atommases`  or masses from a list of masses taken from Wikipedia. The `calculate` method is used for computing the first two terms of the `data.moments` matrix.


`class Moment(Method): `
    `def calculate(self): `
        
        `# Return if moments been parsed already
        if hasattr(self.data, "moments"):
            return True

        # Do we have the needed info in the data object?
        if not hasattr(self.data, "atomcharges"): 
            self.logger.error("Missing atomcharges")
            return False
        if not hasattr(self.data,"atomcoords"):
            self.logger.error("Missing atomcoords")
            return False

        # Should be moved to utils.PeriodicTable later on
        masses = [
            None,
            1.00794, 4.002602,
            6.941, 9.012182, .... 262, 261, 262, 266, 264, 277, 268, 271, 272, 285]

        self.logger.info("Creating attribute moments")
        size = len(self.data.atomnos)
        origin = numpy.zeros(3)

        # Check if atommasses is present, if not use default values to calculate origin
        if not hasattr(self.data, "atommasses"):
            self.logger.info("Missing atommasses, using default masses instead")
            origin = sum(masses[self.data.atomnos[i]] * self.data.atomcoords[-1][i] for i in range(size)) / sum(masses[self.data.atomnos[i] for i in range(size))
        else:
            origin = sum(self.data.atommasses[i] * self.data.atomcoords[-1][i] for i in range (size)) / sum(masses[self.data.atomnos[i] for i in range(size))

        # What to use as chargees, since Lowdin or Mulliken charges cannot be mapped to electronic charges
        dipole = sum(self.data.atomcharges["mulliken"][i] * (self.data.atomcoords[-1][i]-origin) for i in range(size))
        # Hack to set attribute
        self.data.moments = lambda: None
        self.data.moments = [origin, dipole]
        return True #let caller know we finished moments
`

**Comment by ATenderholt** [comment]
What about overloading the constructor to take the desired charges (whether is Lowdin, Mulliken, or something else) and masses? If masses is ```None```, then you can use the default atomic masses.

**Comment by gaursagar** [comment]
Constructor overloading would be a cleaner approach to handle default options, I would make that change.

I think a warning message should be issued while the calculate method is called, to inform users that partial charge schemes 'mulliken' or 'lowdin' are not actual representatives of electronic charge distributions [(ref.)](https://www.researchgate.net/post/Do_Mulliken_atomic_charges_in_the_Gaussian_09_program_have_any_relation_to_the_experimental_ionic_character_of_atoms)

**Comment by langner** [comment]
I think we will end up changing `atmcharges` with #419. Pushing back for now.
